internal/model/model.go

```diff
@@ -254,7 +275,18 @@ func (m *Model) onKeyPressed(key string) (tea.Model, tea.Cmd) {
 			return m, loadFile(adjustedFileName, item.profile)
 		}
 
-		return m, nil
+        return m, nil
+
+    //toggle on and insiate sortByCoverage (default = Asc)
+    case "s":
+        m.toggleSort()
+        m.Update(m.profilesLoaded)
```

Comment:
I think that internal bubbletea methods should not be called by our code; instead, the framework handles these calls externally, and we only need to react to them. Instead, I suggest to implement a new function `updateProfilesList` or similar, which checks the current sort state (type + direction), figures out the items in the correct order, and sets them into the list (`m.list`). 

The resulting code should look something like this:
```go
	case "s":
		m.toggleSort()
		return m.updateListItems()

	case "!":
		m.toggleSortOrder()
		return m.updateListItems()
```

and `updateListItems` will sort the list based on the state, covert the sorted profiles to list items, and call `m.list.setItems`.

Reply:
@orlangure After sorting the list items depending on the SortState, does calling the `m.list.SetItems` automatically update the UI?

Reply:
> After sorting the list items depending on the SortState, does calling the m.list.SetItems automatically update the UI?

Yes, if done properly this will cause the UI to redraw. The flow in bubbletea framework works like "action -> model changes -> rendering", and setting list items is the second step - "model changes". Rendering will follow, and will pick up the updated order of items.

Regarding your second question, below is my draft of the changed functions, based on your code:

```go
func (m *Model) updateListItems() (tea.Model, tea.Cmd) {
	m.sort()

	m.items = make([]list.Item, len(m.profilesLoaded))

	for i, p := range m.profilesLoaded {
		// package name should already be set
		p.FileName = strings.TrimPrefix(p.FileName, m.detectedPackageName+"/")
		m.items[i] = &coverProfile{
			profile:    p,
			percentage: percentCovered(p),
		}
	}

	return m, m.list.SetItems(m.items)
}

func (m *Model) sort() {
	switch m.sortState.Type {
	case sortStateByPercentage:
		m.sortByPercentage()
	case sortStateByName:
		m.sortByName()
	}
}

func (m *Model) sortByPercentage() {
	sort.Slice(m.profilesLoaded, func(i, j int) bool {
		if m.sortState.Order == ASC {
			return percentCovered(m.profilesLoaded[i]) < percentCovered(m.profilesLoaded[j])
		} else {
			return percentCovered(m.profilesLoaded[i]) > percentCovered(m.profilesLoaded[j])
		}
	})
}

func (m *Model) sortByName() {
	sort.Slice(m.profilesLoaded, func(i, j int) bool {
		if m.sortState.Order == ASC {
			return m.profilesLoaded[i].FileName < m.profilesLoaded[j].FileName
		} else {
			return m.profilesLoaded[i].FileName > m.profilesLoaded[j].FileName
		}
	})
}
```

It is not the most efficient and/or elegant way of solving the problem, but it works. It sorts the "loaded profiles" slice, and then loads the already sorted items into the list.

Reply:
Great stuff! I was playing around with this and nearly had this same implementation, just couldn't figure out some of the Golang syntax. Thanks for this assistance. How would you feel about pulling out this code block to have its own function. This same process is currently used twice in the code. Ref `onProfilesLoaded` line 206.
```golang
	m.items = make([]list.Item, len(m.profilesLoaded))

	for i, p := range m.profilesLoaded {
		// package name should already be set
		p.FileName = strings.TrimPrefix(p.FileName, m.detectedPackageName+"/")
		m.items[i] = &coverProfile{
			profile:    p,
			percentage: percentCovered(p),
		}
	}
```

I will push my current changes and open a PR so you can see where I am at. I have created a `model_test.go` file, but I am not too sure where to start with that. 
And then add the button shortcuts to the UI so users will be able to see how to toggle the sort features.

Thanks as always!